### DISCUSSION TITLE
Question about how to self-update without having to manually run the application again.

### DISCUSSION THREAD

**Discussion by LorenzoAlfaro** [dennisvang/notsotuf-example#31] [Q&A]
Hi, is there a way to separate the client from the main app so the client will update the app and then run it after the update? The way the example is set up after an update, the app has to close itself, and then you have to rerun the app manually. I'm just thinking of ways to reduce friction for the end user.

**Comment by dennisvang** [comment]
Hi @LorenzoAlfaro , thanks for your interest in tufup. :-)

It is certainly possible to restart your app after the update has finished.

The way to approach this depends on your platform, and on the way your app is bundled (e.g. if you're using PyInstaller):
- On macOS the app is already restarted, by default, as can be seen [here][1] (although I don't have access to a mac, so cannot verify directly)
- On Windows, this could be done by supplying a custom batch-file template, as described in [tufup #48][5].
  I haven't tested this specific case yet, but here's what the batch template could look like:
  ```python
  custom_batch_template = """@echo off
  {log_lines}
  echo Moving app files...
  robocopy "{src_dir}" "{dst_dir}" {robocopy_options}
  echo Done.
  echo Restarting app
  start "" "{app_exe_path}"
  {delete_self}
  """
  ```
  This adds a line to the [default template][2], using the [windows `start` command][4] to restart the app.
  Note that the template introduces a custom template variable called `app_exe_path`. Both custom variables need to be passed on to the client (also see [tufup #48][5]):
  ```python
  client.download_and_install_update(
      ..., 
      batch_template=custom_batch_template, 
      batch_template_extra_kwargs=dict(app_exe_path=...), 
      ...,
  )
  ```
  The value of `app_exe_path` depends on your bundle. For example, if you're using a PyInstaller bundle, you could use `sys.executable`, or if you're running a plain python script, you could perhaps use something like `f'{sys.executable}\" \"{sys.argv[0]}'`. Have a look at [this explanation][3] as well.

  *As mentioned above, I haven't tested this specific example yet, but I hope the general idea is clear.*

[1]: https://github.com/dennisvang/tufup/blob/4204d68cd33961dce0bf5650db09413845ca9cab/src/tufup/utils/platform_specific.py#L241
[2]: https://github.com/dennisvang/tufup/blob/4204d68cd33961dce0bf5650db09413845ca9cab/src/tufup/utils/platform_specific.py#L88
[3]: https://pyinstaller.org/en/stable/runtime-information.html#using-sys-executable-and-sys-argv-0
[4]: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/start
[5]: https://github.com/dennisvang/tufup/pull/48

**Comment by dennisvang** [comment]
@LorenzoAlfaro Please let us know if you run into any issues. :-)

**Comment by LorenzoAlfaro** [comment]
> 

Hi Dennis, thank you for taking the time for such a complete answer. With your help I managed to make it work!

**Comment by dennisvang** [comment]
Hi @LorenzoAlfaro 
Good to hear it's working.
Thanks for your feedback. :-)

**Comment by RasmusN** [comment]
This is awesome stuff, I that recommend you include it in the tufup documentation.

**Comment by mdhuzaifapatel** [comment]
Hello guys, I have an existing python application could you please guide me how to integrate tufup in that?
also I'm using a Linux as repository to push updates.

so i want to achieve this, I push the new updated application files on server, the old app should fetch new updates from there and update the app. 

please help me out, I'm unable to do it on my own. your help will be appreciated. thanks

**Comment by dennisvang** [comment]
@mdhuzaifapatel see https://github.com/dennisvang/tufup-example/issues/3#issuecomment-1698822223
